include/domain_bridge/service_bridge_impl.inc

```diff
@@ -0,0 +1,202 @@
+ * 
+ * In this case, we need to call another service in our service callbacks, so we need
+ * this beautiful hack :).
+ * 
+ * TODO(ivanpauno): There should be first class support for this in rclcpp.
+ */
+template<typename ServiceT>
+class HackedService
+: public rclcpp::Service<ServiceT>,
+  public std::enable_shared_from_this<HackedService<ServiceT>>
+{
+public:
+  using CallbackT = std::function<void(
+    std::shared_ptr<HackedService<ServiceT>>,
+    const std::shared_ptr<rmw_request_id_t> &,
+    const std::shared_ptr<typename ServiceT::Request> &)>;
+
+  HackedService(
+    std::shared_ptr<rcl_node_t> node_handle,
+    const std::string & service_name,
+    CallbackT my_callback,
+    rcl_service_options_t & service_options)
+  : rclcpp::Service<ServiceT>(node_handle, service_name, rclcpp::AnyServiceCallback<ServiceT>{}, service_options),
+    callback_(std::move(my_callback)) {}
+
+  void
+  handle_request(
+    std::shared_ptr<rmw_request_id_t> request_header,
+    std::shared_ptr<void> request) override
+  {
+    auto typed_request = std::static_pointer_cast<typename ServiceT::Request>(request);
+    // Original implementation:
+    // ***
+    // auto response = std::make_shared<typename ServiceT::Response>();
+    // any_callback_.dispatch(request_header, typed_request, response);
+    // send_response(*request_header, *response);
+    // ***
+    // The hack: forget about sending a response immediately.
+    callback_(this->shared_from_this(), request_header, typed_request);
+  }
+private:
+  // any_callback_ is private in service, we will use our own callback implementation.
+  CallbackT callback_;
+};
+
+/// Create a service with a given type.
+/// \internal
+template<typename ServiceT>
+typename std::shared_ptr<HackedService<ServiceT>>
+create_hacked_service(
+  std::shared_ptr<rclcpp::Node> node,
+  const std::string & service_name,
+  typename HackedService<ServiceT>::CallbackT callback,
+  rclcpp::CallbackGroup::SharedPtr group)
+{
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  auto serv = std::make_shared<HackedService<ServiceT>>(
+    node->get_node_base_interface()->get_shared_rcl_node_handle(),
+    service_name, std::move(callback), service_options);
+  auto serv_base_ptr = std::static_pointer_cast<rclcpp::ServiceBase>(serv);
+  node->get_node_services_interface()->add_service(serv_base_ptr, group);
+  return serv;
+}
+
+rclcpp::Node::SharedPtr
+get_node_for_domain(DomainBridgeImpl & impl, std::size_t domain_id);
+
+const std::string &
+get_node_name(const DomainBridgeImpl & impl);
+}  // namespace detail
+
+/// Bridge a service from one domain to another.
+/**
+ * \param service_bridge: Struct containing info about the service to bridge.
+ * \param options: Options for bridging the topic.
+ */
+template<typename ServiceT>
+void
+DomainBridge::bridge_service(
+  const TopicBridge & service_bridge,
+  const TopicBridgeOptions & options)
```

Comment:
nitpick: reusing the "topic" bridge types looks odd. It would look more semantically correct if they were named "ServiceBridge" and "ServiceBridgeOptions" respectively, even if they are just aliases to the topic types.

Reply:
Agreed, I will fix that.

Reply:
Fixed in https://github.com/ros2/domain_bridge/pull/26/commits/113aa815239703dea2652306b4e0fcb885834ca5.

Now I'm imitating

https://github.com/ros2/domain_bridge/blob/113aa815239703dea2652306b4e0fcb885834ca5/include/domain_bridge/domain_bridge.hpp#L120-L126

signature.

I'm still using `TopicBridge` in the implementation, but it's not actually part of the API so I think it doesn't matter for the moment.

---

I've enforced the [rule of zero](https://en.cppreference.com/w/cpp/language/rule_of_three) in the `ServiceBridgeOptions` class, and also in the current `TopicBridgeOptions` and `QosOptions` classes.

Reply:
> I'm still using TopicBridge in the implementation

Fixed that in https://github.com/ros2/domain_bridge/pull/26/commits/752d196acab2ac71a9b8618a6a95cc24f59b181d.